File: languages/js/src/builder.ts
import { BloockBridge } from "./bridge/bridge";
import {
  RecordTypes,
  RecordBuilderFromStringRequest,
  RecordBuilderFromJSONRequest,
  RecordBuilderFromHexRequest,
  RecordBuilderFromBytesRequest,
  RecordBuilderFromFileRequest,
  RecordBuilderFromRecordRequest,
  Signer as SignerProto,
  Encrypter as EncrypterProto,
  Decrypter as DecrypterProto,
  RecordBuilderFromLoaderRequest
} from "./bridge/proto/record";

import { Record } from "./entity/record";
import { Signer } from "./entity/signer";
import { Encrypter } from "./entity/encrypter";
import { Decrypter } from "./entity/decrypter";
import { Loader } from "./entity/loader";
import { NewConfigData } from "./config/config";

export class RecordBuilder {
  payload: any;
  payloadType!: RecordTypes;
  signer: SignerProto | undefined;
  encrypter: EncrypterProto | undefined;
  decrypter: DecrypterProto | undefined;

  private constructor(payload: any, payloadType: RecordTypes) {
    this.payload = payload;
    this.payloadType = payloadType;
  }

  public static fromString(str: string): RecordBuilder {
    return new RecordBuilder(str, RecordTypes.STRING);
  }

  public static fromJson(json: any): RecordBuilder {
    return new RecordBuilder(JSON.stringify(json), RecordTypes.JSON);
  }

  public static fromHex(hex: string): RecordBuilder {
    return new RecordBuilder(hex, RecordTypes.HEX);
  }

  public static fromBytes(bytes: Uint8Array): RecordBuilder {
    return new RecordBuilder(bytes, RecordTypes.BYTES);
  }

  public static fromFile(bytes: Uint8Array): RecordBuilder {
    return new RecordBuilder(bytes, RecordTypes.FILE);
  }

  public static fromRecord(bytes: Record): RecordBuilder {
    return new RecordBuilder(bytes, RecordTypes.RECORD);
  }

  public static fromLoader(loader: Loader): RecordBuilder {
    return new RecordBuilder(loader, RecordTypes.LOADER);
  }

  public withSigner(signer: Signer): RecordBuilder {
    this.signer = signer.toProto();
    return this;
  }

  public withEncrypter(encrypter: Encrypter): RecordBuilder {
    this.encrypter = encrypter.toProto();
    return this;
  }

  public withDecrypter(decrypter: Decrypter): RecordBuilder {
    this.decrypter = decrypter.toProto();
    return this;
  }

  async build(): Promise<Record> {
    const bridge = new BloockBridge();
    switch (this.payloadType) {
      case RecordTypes.STRING: {
        const req = RecordBuilderFromStringRequest.fromPartial({
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromString(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
      case RecordTypes.JSON: {
        const req = RecordBuilderFromJSONRequest.fromPartial({
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromJson(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
      case RecordTypes.HEX: {
        const req = RecordBuilderFromHexRequest.fromPartial({
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromHex(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
      case RecordTypes.BYTES: {
        const req = RecordBuilderFromBytesRequest.fromPartial({
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromBytes(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
      case RecordTypes.FILE: {
        const req = RecordBuilderFromFileRequest.fromPartial({
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromFile(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
      case RecordTypes.RECORD: {
        const req = RecordBuilderFromRecordRequest.fromPartial({
          payload: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromRecord(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
      case RecordTypes.LOADER: {
        const req = RecordBuilderFromLoaderRequest.fromPartial({
          configData: NewConfigData(),
          loader: this.payload,
          signer: this.signer,
          encrypter: this.encrypter,
          decrypter: this.decrypter
        });

        return bridge
          .getRecord()
          .BuildRecordFromLoader(req)
          .then(res => {
            if (res.error) {
              throw res.error;
            }
            return Record.fromProto(res.record!);
          });
      }
    }
    return Promise.reject(new Error("Unexpected record type"));
  }
}

File: languages/python/bloock/client/builder.py
from __future__ import annotations
from bloock._bridge import bridge
import bloock._bridge.proto.record_pb2 as proto
from bloock._bridge.proto.shared_pb2 import Error
from bloock.client.entity.encrypter import Encrypter
from bloock.client.entity.decrypter import Decrypter
from bloock.client.entity.record import Record
from bloock.client.entity.signer import Signer
from bloock.client.entity.loader import Loader
from bloock._config.config import Config


class RecordBuilder:
    def __init__(
        self,
        payload,
        payload_type: proto.RecordTypes.ValueType,
        signer: proto.Signer | None,
        encrypter: proto.Encrypter | None,
        decrypter: proto.Decrypter | None,
    ) -> None:
        self.payload = payload
        self.payload_type = payload_type
        self.signer = signer
        self.encrypter = encrypter
        self.decrypter = decrypter

    @staticmethod
    def from_record(record: Record) -> RecordBuilder:
        return RecordBuilder(
            payload=record.to_proto(),
            payload_type=proto.RecordTypes.RECORD,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    @staticmethod
    def from_loader(loader: Loader) -> RecordBuilder:
        return RecordBuilder(
            payload=loader.to_proto(),
            payload_type=proto.RecordTypes.LOADER,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    @staticmethod
    def from_string(string: str) -> RecordBuilder:
        return RecordBuilder(
            payload=string,
            payload_type=proto.RecordTypes.STRING,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    @staticmethod
    def from_hex(hex: str) -> RecordBuilder:
        return RecordBuilder(
            payload=hex,
            payload_type=proto.RecordTypes.HEX,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    @staticmethod
    def from_json(json: str) -> RecordBuilder:
        return RecordBuilder(
            payload=json,
            payload_type=proto.RecordTypes.JSON,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    @staticmethod
    def from_file(file: bytes) -> RecordBuilder:
        return RecordBuilder(
            payload=file,
            payload_type=proto.RecordTypes.FILE,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    @staticmethod
    def from_bytes(b: bytes) -> RecordBuilder:
        return RecordBuilder(
            payload=b,
            payload_type=proto.RecordTypes.BYTES,
            signer=None,
            encrypter=None,
            decrypter=None,
        )

    def with_signer(self, signer: Signer) -> RecordBuilder:
        self.signer = signer.to_proto()
        return self

    def with_encrypter(self, encrypter: Encrypter) -> RecordBuilder:
        self.encrypter = encrypter.to_proto()
        return self

    def with_decrypter(self, decrypter: Decrypter) -> RecordBuilder:
        self.decrypter = decrypter.to_proto()
        return self

    def build(self) -> Record:
        client = bridge.BloockBridge()

        res = proto.RecordBuilderResponse()

        if self.payload_type == proto.RecordTypes.RECORD:
            res = client.record().BuildRecordFromRecord(
                proto.RecordBuilderFromRecordRequest(
                    config_data=Config.new(),
                    payload=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )
        elif self.payload_type == proto.RecordTypes.LOADER:
            res = client.record().BuildRecordFromLoader(
                proto.RecordBuilderFromLoaderRequest(
                    config_data=Config.new(),
                    loader=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )
        elif self.payload_type == proto.RecordTypes.STRING:
            res = client.record().BuildRecordFromString(
                proto.RecordBuilderFromStringRequest(
                    config_data=Config.new(),
                    payload=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )
        elif self.payload_type == proto.RecordTypes.HEX:
            res = client.record().BuildRecordFromHex(
                proto.RecordBuilderFromHexRequest(
                    config_data=Config.new(),
                    payload=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )
        elif self.payload_type == proto.RecordTypes.JSON:
            res = client.record().BuildRecordFromJson(
                proto.RecordBuilderFromJSONRequest(
                    config_data=Config.new(),
                    payload=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )
        elif self.payload_type == proto.RecordTypes.FILE:
            res = client.record().BuildRecordFromFile(
                proto.RecordBuilderFromFileRequest(
                    config_data=Config.new(),
                    payload=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )
        elif self.payload_type == proto.RecordTypes.BYTES:
            res = client.record().BuildRecordFromBytes(
                proto.RecordBuilderFromBytesRequest(
                    config_data=Config.new(),
                    payload=self.payload,
                    signer=self.signer,
                    encrypter=self.encrypter,
                    decrypter=self.decrypter,
                )
            )

        if res.error != Error():
            raise Exception(res.error.message)

        return Record.from_proto(res.record)
